src/libUtils/DnsUtils.cpp

```diff
@@ -0,0 +1,329 @@
+    return false;
+  }
+
+  ns_rr rr;
+  ns_parserr(&nsMsg, ns_s_an, 0, &rr);
+  auto rrData = ns_rr_rdata(rr) + 1;
+
+  string p(rrData, rrData + rr.rdlength - 1);
+
+  if (p.compare("0") == 0 || p.empty()) {
+    LOG_GENERAL(WARNING, "Pubkey from dns is 0 or empty: " << pubKeyUrl);
+    return false;
+  }
+
+  if (!DataConversion::HexStrToUint8Vec(p, output)) {
+    LOG_GENERAL(WARNING, "Invalid data obtained from pub key " << pubKeyUrl);
+    return false;
+  }
+  return true;
+}
+
+void QueryDnsList(DnsListType listType) {
+  LOG_MARKER();
+  if (isShuttingDown) return;
+
+  auto url = dnsAddresses[listType];
+  if (url.empty()) {
+    LOG_GENERAL(INFO, "DNS address is empty for type " << (int)listType);
+    return;
+  }
+
+  auto &dataListCache = dnsCacheListDataMap[listType];
+
+  if (!dataListCache.dataAccessMutex.try_lock()) {
+    LOG_GENERAL(INFO, "Another thread is querying " << dnsAddresses[listType]);
+    return;
+  }
+
+  if (!QueryIpStrListFromDns(listType)) {
+    LOG_GENERAL(WARNING, "Failed to obtain ip list from "
+                             << dnsAddresses[listType]
+                             << ", try again on another DS epoch");
+    dataListCache.dataAccessMutex.unlock();
+    return;
+  }
+
+  vector<uint128_t> currentIpKeys;
+  currentIpKeys.reserve(dataListCache.ipList.size());
+
+  auto &pubKeyList = dataListCache.pubKeyList;
+
+  // Adding new pubkeys to our dns cache
+  for (const auto &ipStr : dataListCache.ipList) {
+    uint128_t ipKey;
+    if (!IPConverter::ToNumericalIPFromStr(ipStr, ipKey)) {
+      LOG_GENERAL(WARNING, "Unable to change IP to ipKey: " << ipKey);
+      continue;
+    }
+
+    currentIpKeys.emplace_back(ipKey);
+
+    if (pubKeyList.find(ipKey) != pubKeyList.end()) {
+      // Already exists, don't need to query again, unlikely to change pubkey
+      // for an ip
+      continue;
+    }
+
+    auto pubKeyUrl = GetPubKeyUrl(ipStr, url);
+    bytes pubKeyResult;
+    if (QueryPubKeyFromUrl(pubKeyResult, pubKeyUrl)) {
+      pubKeyList[ipKey] = pubKeyResult;
+    }
+  }
+
+  // Remove pubkeys that are no longer in the dns list
+  for (auto itr = pubKeyList.begin(); itr != pubKeyList.end();) {
+    if (find(currentIpKeys.begin(), currentIpKeys.end(), itr->first) ==
+        currentIpKeys.end()) {
+      itr = pubKeyList.erase(itr);
+    } else {
+      ++itr;
+    }
+  }
+
+  dataListCache.dataAccessMutex.unlock();
+}
+}  // namespace
+
+void InitDnsCacheList() {
+  dnsCacheListDataMap[DnsListType::UPPER_SEED];
+  dnsCacheListDataMap[DnsListType::L2L_DATA_PROVIDERS];
+  dnsCacheListDataMap[DnsListType::MULTIPLIERS];
+
+  dnsAddresses[DnsListType::UPPER_SEED] = UPPER_SEED_DNS;
+  dnsAddresses[DnsListType::L2L_DATA_PROVIDERS] = L2L_DATA_PROVIDERS_DNS;
+  dnsAddresses[DnsListType::MULTIPLIERS] = MULTIPLIER_DNS;
+
+  isShuttingDown = false;
+}
+
+void ShutDownDnsCacheList() { isShuttingDown = true; }
+
+void AttemptPopulateLookupsDnsCache() {
```

Comment:
Is DNS timeout resolution removed for some reason? 
When the `QueryDnsList`  is called for the first DNS resolution, it locks the mutex and call `QueryIpStrListFromDns` and unlocks the mutex at the last. If there's no timeout, it may be possible that the calling thread blocks at the call to `QueryIpStrListFromDns`, and this may result in other `QueryDnsList` not getting called.

Reply:
The `try_lock()` won't be blocked, it will check if the current thread cannot acquire the lock, it will return false.

In that case, we will fall back to the local list and continue to progress.

Previous timeout isn't really a timeout for getaddrinfo, but just a timeout to move on to local list, while getaddrinfo is still being blocked.

With this implementation, there's 1 lesser thread required.

getaddrinfo can block indefinitely, the failing cases i saw can block up to 15 seconds so far.

Instead of spawning more threads to query when DNS is down, I think it can wait for getaddrinfo to end, and retry again.